src/aws-cpp-sdk-core/source/http/curl-multi/CurlMultiHttpClient.cpp

```diff
@@ -0,0 +1,833 @@
+
+    for(Curl::CurlEasyHandleContext* handleCtx : m_multiTasks)
+    {
+        curl_multi_remove_handle(m_curlMultiHandleContainer.AccessCurlMultiHandle(), handleCtx->m_curlEasyHandle);
+        handleCtx->curlResultMsg = nullptr;
+        handleCtx->curlResult = static_cast<CURLcode>(-1);
+        handleCtx->m_onCurlDoneFn();
+        m_multiTasks.erase(handleCtx);
+    }
+    m_multiTasks.clear();
+}
+
+void CurlMultiHttpClient::CurlMultiPerformThread(CurlMultiHttpClient* pClient)
+{
+    assert(pClient && pClient->m_curlMultiHandleContainer.AccessCurlMultiHandle());
+    CURLM *multi_handle = pClient->m_curlMultiHandleContainer.AccessCurlMultiHandle();
+
+    int stillRunning = 0;
+    while(pClient->m_isRunning)
+    {
+        // wait for new task signal
+        {
+            std::unique_lock<std::mutex> lockGuard(pClient->m_signalMutex);
+            pClient->m_signalRunning.wait(lockGuard,
+                                          [pClient, &stillRunning]
+                                          {
+                                            // return true to unlock
+                                            if(!pClient->m_isRunning.load())
+                                              return true;
+
+                                            return pClient->m_tasksQueued || stillRunning;
+                                          });
+        }
+        if(!pClient->m_isRunning.load())
+        {
+            break;
+        }
+
+        pClient->m_tasksQueued = 0;
+
+        CURLMcode mc = curl_multi_perform(multi_handle, &stillRunning);
+        if(mc != CURLM_OK)
+        {
+          AWS_LOGSTREAM_ERROR(CURL_HTTP_CLIENT_TAG, "Curl curl_multi_perform returned error code " << mc
+                                                << " resetting multi handle.");
+          pClient->CurlMultiPerformReset();
+        }
+        int msgQueue = 0;
+        do {
+
+          struct CURLMsg* message = curl_multi_info_read(multi_handle, &msgQueue);
+          if(message)
+          {
+            if(message->msg == CURLMSG_DONE)
+            {
+              CURL* easyHandle = message->easy_handle;
+              assert(easyHandle);
+              Curl::CurlEasyHandleContext* pEasyHandleCtx = nullptr;
+              curl_easy_getinfo(easyHandle, CURLINFO_PRIVATE, &pEasyHandleCtx);
+              assert(pEasyHandleCtx);
+
+              pEasyHandleCtx->curlResult = message->data.result;
+              pEasyHandleCtx->m_onCurlDoneFn();
+            } else {
+              assert(!"Todo");
+            }
+          }
+        } while(msgQueue > 0);
+
+        if(!mc && stillRunning)
+        {
+            /* wait for activity, timeout or "nothing" */
+            mc = curl_multi_poll(multi_handle, NULL, 0, 1000, NULL);
+            if(mc)
+            {
+              AWS_LOGSTREAM_ERROR(CURL_HTTP_CLIENT_TAG, "Curl curl_multi_poll returned error code " << mc);
+              break;
```

Comment:
this will kill the curl thread? is an error here fatal? what happens to user code if thread exits here